Java/core/src/main/java/com/amazon/randomcutforest/PredictiveRandomCutForest.java

```diff
@@ -0,0 +1,565 @@
+                checkArgument(missingValues[i] >= 0, " missing values cannot be at negative position");
+                checkArgument(missingValues[i] <= inputPoint.length,
+                        "missing values cannot be at position larger than input length");
+                checkArgument(forestMode == ForestMode.TIME_AUGMENTED || missingValues[i] < inputPoint.length,
+                        "cannot be equal to input length");
+                newMissingValues[i] = (forestMode == ForestMode.STREAMING_IMPUTE) ? startPosition + missingValues[i]
+                        : missingValues[i];
+            }
+        }
+        // check when TIME_AUGMENTED and missingValue includes timestamp
+        float[] point = preprocessor.getScaledShingledInput(toDoubleArray(inputPoint), timestamp, missingValues,
+                forest);
+        if (point == null) {
+            return new SampleSummary(preprocessor.getInputLength());
+        }
+        return preprocessor.invertInPlaceRecentSummaryBlock(forest.getConditionalFieldSummary(point, newMissingValues,
+                numberOfRepresentatives, shrinkage, true, false, centrality, preprocessor.getShingleSize()));
+    }
+
+    public SampleSummary predict(float[] inputPoint, long timestamp, int[] missingValues) {
+        return predict(inputPoint, timestamp, missingValues, 5, 0.3, 0.5);
+    }
+
+    /**
+     * Near neighbors is an extreme example of prediction where there aare no
+     * missing values
+     * 
+     * @param inputPoint the input
+     * @param timestamp  the timestamp (not used unless for TIME_AUGMENTED and
+     *                   STREAMING_IMPUTE)
+     * @return returns a clustering of the near neighbors in the (time decaying)
+     *         sample maintained by RCF
+     */
+    public SampleSummary neighborSummary(float[] inputPoint, long timestamp) {
+        return predict(inputPoint, timestamp, null, 1, 0, 1);
+    }
+
+    /**
+     * This is the anomalyScore() function in RCF -- that definition of score has
+     * diverged from the original paper and perhaps the more descriptive name is
+     * more useful. Note that it is important that this is Expectation of Inverse
+     * versus Inverse of Expectation -- a la Jensen's Inequality.
+     * 
+     * @param inputPoint the input point
+     * @param timestamp  the timestamp
+     * @return a score in the range (0:log_2(sampleSize)) where larger values
+     *         indicate potential anomalies. Note that the score can remain elevated
+     *         due to shingling for a while -- use
+     *         ParkServices/ThresholdedRandomCutForest where the PredictorCorrector
+     *         architecture is used to evaluate the score further and produce a more
+     *         refined anomalyGrade()in [0:1] ; including setting the grade to 0
+     *         (corrector) even though the score (of this function) is high.
+     */
+    public double getExpectedInverseDepthScore(float[] inputPoint, long timestamp) {
```

Comment:
ThresholdedRandomCutForest (TRCF, for now) is geared towards time series -- it uses the predictor-corrector to flag anomalies, and uses the single interface process() that scores as well as updates -- which is significantly more efficient. For time series AD, TRCF makes sense (as is being used in OpenSearch at the moment I believe). For forecasting, the extension of TRCF, RCFCaster would be natural.


However there has always been questions about more control regarding updates and scoring -- not all such usages are clearly reasonable. For example updating on point X; and then scoring point X creates a dependency that is not information theoretically clean (or explainable, etc., because X can be a multitude of points and some were seen before, some were not etc). But the question is : should the controls exist (under the generous assumption that the calling function knows what it is doing)? It seems so -- and so if there is a desire to intersperse updates and scores, then this class would allow it -- whereas TRCF would not (and force the calling function to respect arrow of time/seen vs unseen data).